### ISSUE TITLE
docs: Ambiguity / shadowing in doc example

### ISSUE THREAD

**Issue by tony** [clokep/celery-batches#42]
Hi!  Thank you for the project @clokep. 

I have some confusion re: the docs example, between the conventions:

[`requests`](https://pypi.org/project/requests/) vs [celery](https://docs.celeryproject.org/)'s [`celery.worker.request`](https://docs.celeryproject.org/en/v4.4.7/reference/celery.worker.request.html) vs celery-batches's own [`celery_batches.SimpleRequest`](https://celery-batches.readthedocs.io/en/latest/?badge=latest#celery_batches.SimpleRequest) vs the example's `requests` param in `wot_api(requests)`

```python
import requests
from urlparse import urlparse

from celery_batches import Batches

wot_api_target = 'https://api.mywot.com/0.4/public_link_json'

@app.task(base=Batches, flush_every=100, flush_interval=10)
def wot_api(requests):
    sig = lambda url: url
    reponses = wot_api_real(
        (sig(*request.args, **request.kwargs) for request in requests)
    )
    # use mark_as_done to manually return response data
    for response, request in zip(reponses, requests):
        app.backend.mark_as_done(request.id, response, request=request)


def wot_api_real(urls):
    domains = [urlparse(url).netloc for url in urls]
    response = requests.get(
        wot_api_target,
        params={'hosts': ('/').join(set(domains)) + '/'}
    )
    return [response.json[domain] for domain in domains]
```

`requests` is imported, but it also seems to be used in `def wot_api(requests)`.

Proposal:

To avoid ambiguity, we can make an example that avoids using the `requests` package completely (despite it being ubiquitous, oddly enough its naming collides in this example 😆)

@clokep Are you open to collaborating with me on improving the docs? I'm willing to spend a few hours cleaning up (but not sure if you have your own ideas in mind)

**Comment by tony** [comment]
I will look at this evening and during weekend / early next week

General sphinx question:

What do you think of @pradyunsg's [furo](https://github.com/pradyunsg/furo) theme?  ([example](https://pradyunsg.me/furo/quickstart/))

Aside: I'm thinking of using it on my sites as a replacement for the flask-like theme (alabaster).

**Comment by pradyunsg** [comment]
A few other examples:

- https://pip.pypa.io/en/stable/
- https://black.readthedocs.io/en/stable/
- https://urllib3.readthedocs.io/en/stable/

;)

**Comment by tony** [comment]
Considering the code example above, is there anything you can do in the here and now to distinguish the variables?  e.g. use `urllib.request`? 

> This example also doesn't really show _why_ you would use celery-batches (the example could just as easily be a single task).

Yes, perhaps it'd be worth writing a few more? Actually maybe you would be best to do it since you show permitted uses that are used to help validate the reader understands correctly

**Comment by tony** [comment]
@clokep Do you think you would be best fit for taking an initial stab at potentially tweaking the docs and I check if I have any ideas after? Or should I propose a PR this weekend?

The reason why is I think you would be in the best position to give clear examples and I think it's easy enough to configure sphinx.  (viewing the source of pip, black, and urllib's own sphinx docs may be good inspiration on the best practices too)

**Comment by clokep** [comment]
> What do you think of @pradyunsg's [furo](https://github.com/pradyunsg/furo) theme? ([example](https://pradyunsg.me/furo/quickstart/))

It seems fine, I think anything that's clean is good. I'm more concerned about the content. (I think I've usually just use the alabaster because it is one less thing that might go unmaintained.)

> @clokep Do you think you would be best fit for taking an initial stab at potentially tweaking the docs and I check if I have any ideas after? Or should I propose a PR this weekend?

I do not think I'll have time this weekend to write anything up. If you have time that would be great, if not if you could at least detail some of the confusion / pain points that would be good.

In terms of better examples, some of the things I have found batchig of tasks to be useful for include:

* De-duplicate work.
* Handle accumulating data.
* Kick off algorithms that have a high start-up cost (and lower incremental cost per item).

... I'm sure there are others too though and would be curious why others are using it!

**Comment by tony** [comment]
> It seems fine, I think anything that's clean is good. I'm more concerned about the content. (I think I've usually just use the alabaster because it is one less thing that might go unmaintained.)


Should I try setting `furo` for kicks so we can see how it looks?


> In terms of better examples, some of the things I have found batchig of tasks to be useful for include:
> 
> * De-duplicate work.
> * Handle accumulating data.
> * Kick off algorithms that have a high start-up cost (and lower incremental cost per item).
> 
> ... I'm sure there are others too though and would be curious why others are using it!

If you could write code examples of correct usage, even if it was pseudocode and refer to a magic function to do the deduplication / accumulating / initial expensive warm up, that would be helpful (so long as the celery / celery-batches code was correct)

It would be a major jump start / boost for me to have those, even if you were to jot them down in a comment.

> I do not think I'll have time this weekend to write anything up. If you have time that would be great, if not if you could at least detail some of the confusion / pain points that would be good.

I think for superstar docs, the core will be revolve around examples / usage patterns.  Essential as the autodoc API docs IMO, especially since this is scheduler related

I think what I provided would be merely decorating/bikeshedding - but I'll go far as I can over the weekend

**Comment by clokep** [comment]
> If you could write code examples of correct usage, even if it was pseudocode and refer to a magic function to do the deduplication / accumulating / initial expensive warm up, that would be helpful (so long as the celery / celery-batches code was correct)

I threw together a few ideas below, I'm not sure if they're super helpful or not. The last two touch a bit on using Django (or really any other ORM or datastore you might want). They're a bit contrived, and I didn't include the calling code to send the tasks, but hopefully that's the straightforward part since it isn't different than Celery?

## Expensive task

```python
from collections import defaultdict
from somewhere import expensive_processor

@app.task(base=Batches, flush_every=100, flush_interval=10)
def expensive_task(requests):
    """De-duplicate incoming arguments to only do expensive task once per unique inputs."""
    # Generate a map of unique args -> requests.
    calls = defaultdict(list)
    for request in requests:
        calls[request.args].append(request)

    # Do the expensive processing.
    for call_args, call_requests in calls.items():
		result = expensive_processor(*args)

	    # use mark_as_done to manually return response data
    	for request in call_requests:
        	app.backend.mark_as_done(request.id, result, request=request)
```

## Accumulate

This is useful if you want to reduce some database queries (and some error / potential for dropping data is OK). Maybe if you're keeping some statistics, but not things you'd be billing on, etc.

```python
@app.task(base=Batches, flush_every=100, flush_interval=10)
def last_seen(requests):
    """De-duplicate incoming arguments to only do expensive task once per unique inputs."""
    # Generate a map of unique args -> requests.
    last_seen = {}
    for request in requests:
        user_id, when = request.args
		if user_id not in last_seen or last_seen[user_id] < when:
			last_seen[user_id] = when

	# Update your datastore.
	for user_id, when in last_seen.items():
		User.objects.filter(id=user_id).update(last_logged_in=when)
```

## Bulk updating

It can also be useful to just bulk insert data as quickly as possible, but when the discrete data is from separate tasks.

```python
@app.task(base=Batches, flush_every=100, flush_interval=10)
def bulk_insert(requests):
    """De-duplicate incoming arguments to only do expensive task once per unique inputs."""
    data = []
    for request in requests:
        data.append(MyModel(request.args))

	MyModel.objects.bulk_create(data)
```

**Comment by tony** [comment]
@clokep I missed this! I will add to #45

**Comment by tony** [comment]
I think we should investigate merging #45 as-is without the examples.

If that's okay, I propose we make examples that have tests so we can get a heads up if they break:

https://cihai.git-pull.com/examples.html ([example .py files](https://github.com/cihai/cihai/tree/v0.12.0/examples) through `.. include::`, [examples.md](https://github.com/cihai/cihai/blob/v0.12.0/docs/examples.md))

Tests for doc examples: https://github.com/cihai/cihai/blob/v0.12.0/tests/test_examples.py

**Comment by tony** [comment]
Fixed via 06df39e from https://github.com/clokep/celery-batches/pull/49
